lib/TPP/ConvertPackUnpackOptimization.cpp

```diff
@@ -0,0 +1,189 @@
+//
+
+#include "TPP/BuilderUtils.h"
+#include "TPP/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+using namespace mlir::tpp;
+
+#define GEN_PASS_CLASSES
+#include "TPP/Passes.h.inc"
+namespace {
+
+template <typename Op>
+struct ConvertPackUnpackOptimizationOp : public OpRewritePattern<Op> {
+  using OpRewritePattern<Op>::OpRewritePattern;
+  LogicalResult matchAndRewrite(Op packOp,
+                                PatternRewriter &rewriter) const override {
+    if (packOp.getStaticInnerTiles().size() <= 0)
+      return failure();
+
+    constexpr bool IsPack =
+        std::is_same<mlir::tensor::PackOp, Op>::value == true;
+    ArrayRef<int64_t> shape = IsPack ? packOp.getSourceType().getShape()
+                                     : packOp.getDestType().getShape();
+    int numLoops = shape.size();
+
+    for (size_t i = 0; i < packOp.getInnerDimsPos().size() - 1; i++) {
+      if (packOp.getInnerDimsPos()[i] >= packOp.getInnerDimsPos()[i + 1]) {
+        return failure();
+      }
+    }
+
+    for (size_t i = 0; i < packOp.getStaticInnerTiles().size(); i++) {
+      if (shape[packOp.getInnerDimsPos()[i]] %
+              packOp.getStaticInnerTiles()[i] !=
+          0) {
+        return failure();
+      }
+    }
+    auto zero = getConstIndex(rewriter, 0);
+    auto one = getConstIndex(rewriter, 1);
+
+    SmallVector<Value> lbs(numLoops, zero);
+    SmallVector<Value> ubs;
+    SmallVector<Value> steps(numLoops, one);
+
+    std::map<int, int> tiledDims;
+    for (auto tiledDim : llvm::enumerate(packOp.getInnerDimsPos())) {
+      tiledDims[tiledDim.value()] = tiledDim.index();
+    }
+
+    for (int i = 0; i < numLoops; i++) {
+      if (tiledDims.count(i) && shape[i] != ShapedType::kDynamic &&
+          packOp.getStaticInnerTiles()[tiledDims[i]] != ShapedType::kDynamic) {
+        ubs.push_back(getConstIndex(
+            rewriter, shape[i] / packOp.getStaticInnerTiles()[tiledDims[i]]));
+
+      } else {
+        ubs.push_back(getConstIndex(rewriter, shape[i]));
+      }
+    }
+
+    auto loopNest = mlir::scf::buildLoopNest(
+        rewriter, packOp.getLoc(), lbs, ubs, steps, packOp.getDest(),
+        [&packOp, &numLoops, &tiledDims,
+         &IsPack](OpBuilder &rewriter, Location loc, ValueRange localIvs,
+                  ValueRange iterArgs) -> scf::ValueVector {
+          SmallVector<OpFoldResult> extractSliceOffsets;
+          SmallVector<OpFoldResult> extractSliceStrides;
+          SmallVector<OpFoldResult> extractSliceSizes;
+          SmallVector<OpFoldResult> insertSliceOffsets;
+          SmallVector<OpFoldResult> insertSliceStrides;
+          SmallVector<OpFoldResult> insertSliceSizes;
+
+          auto setArgs = [&numLoops, &tiledDims, &rewriter, &localIvs, &packOp,
```

Comment:
This looks dangerous. Not only the lambda is overriding the same variables from the main function, but it's doing so in the opposite meaning for one case and not the other. People are going to get confused over this code.

I suggest you rename it to something like `sourceSlice*` and `destSlice*`. Also, please add a comment before creating the lambda to understand its construction.